WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,765 @@
+               }
+#endif /* D_COMRV_OVL_DATA_SUPPORT */
+               stComrvCB.stOverlayCache[ucIndex].unProperties = pEntry->unProperties;
+               stComrvCB.stOverlayCache[ucIndex].unToken.uiValue = pEntry->unToken.uiValue;
+#ifdef D_COMRV_EVICTION_LRU
+               stComrvCB.stOverlayCache[pEntry->ucPrevIndex].ucNextIndex = ucIndex;
+               stComrvCB.stOverlayCache[pEntry->ucNextIndex].ucPrevIndex = ucIndex;
+#elif defined(D_COMRV_EVICTION_LFU)
+#elif defined(D_COMRV_EVICTION_MIX_LRU_LFU)
+#endif /* D_COMRV_EVICTION_LRU */
+            }
+            ucEntryIndex++;
+         }
+         ucNumOfEvictionCandidates++;
+      }
+      /* this means the end user locked all entries */
+      else if (ucNumOfEvictionCandidates == 0)
+      {
+         M_COMRV_EXIT_CRITICAL_SECTION();
+         comrvNotificationHook(D_COMRV_NO_AVAILABLE_ENTRY_ERR, unToken.uiValue);
+      }
+      ucIndex = ucEvictCandidateList[ucEntryIndex];
+      /* update the entry access */
+      comrvUpdateHeapEntryAccess(ucIndex);
+      /* update the heap entry with the new token */
+      stComrvCB.stOverlayCache[ucIndex].unToken.uiValue = unToken.uiValue;
+      /* update the heap entry properties with the group size */
+      stComrvCB.stOverlayCache[ucIndex].unProperties.stFields.ucSizeInMinGroupSizeUnits = usOverlayGroupSize;
+      /* if evicted size is larger than requested size we need to update the remaining tail*/
+      ucSizeOfEvictionCandidates = ucEvictCandidateList[ucNumOfEvictionCandidates] - usOverlayGroupSize;
+      if (ucSizeOfEvictionCandidates != 0)
+      {
+         ucEntryIndex += usOverlayGroupSize;
+         stComrvCB.stOverlayCache[ucEntryIndex].unProperties.stFields.ucSizeInMinGroupSizeUnits = ucSizeOfEvictionCandidates;
+#ifdef D_COMRV_EVICTION_LRU
+         stComrvCB.stOverlayCache[stComrvCB.stOverlayCache[ucEntryIndex].ucPrevIndex].ucNextIndex = stComrvCB.stOverlayCache[ucEntryIndex].ucNextIndex;
+         stComrvCB.stOverlayCache[ucEntryIndex].ucNextIndex           = ucIndex;
+         stComrvCB.stOverlayCache[ucEntryIndex].ucPrevIndex           = D_COMRV_LRU_FIRST_ITEM;
+         /* mark the group ID so that it won't pop in the next search */
+         stComrvCB.stOverlayCache[ucEntryIndex].unToken.uiValue       = D_COMRV_ENTRY_TOKEN_INIT_VALUE;
+         stComrvCB.stOverlayCache[ucEntryIndex].unProperties.ucValue &= D_COMRV_ENTRY_PROPERTIES_RESET_MASK;
+         stComrvCB.ucLruIndex = ucIndex;
+#elif defined(D_COMRV_EVICTION_LFU)
+#elif defined(D_COMRV_EVICTION_MIX_LRU_LFU)
+#endif /* D_COMRV_EVICTION_LRU */
+      }
+      // TODO: protect the soon loaded ram
+      /* it is safe now to get new requests */
+      M_COMRV_EXIT_CRITICAL_SECTION();
+      /* the group size in bytes */
+      usOverlayGroupSize <<= 9;
+      /* now we can load the overlay group */
+      pAddress = comrvLoadOvlayGroupHook(M_COMRV_GET_GROUP_OFFSET_IN_BYTES(stComrvCB.stOverlayCache[ucIndex].unToken),
```

Comment:
replace functions arguments with structure

Reply:
done in ae5bf40